contracts/cw721-base/src/contract.rs

```diff
@@ -14,442 +12,609 @@ use cw721::{
 
 use crate::error::ContractError;
 use crate::msg::{ExecuteMsg, InstantiateMsg, MintMsg, MinterResponse, QueryMsg};
-use crate::state::{
-    increment_tokens, num_tokens, tokens, Approval, TokenInfo, CONTRACT_INFO, MINTER, OPERATORS,
-};
-use cw_storage_plus::Bound;
+use crate::state::{token_owner_idx, Approval, TokenIndexes, TokenInfo};
+use cw_storage_plus::{Bound, IndexedMap, Item, Map, MultiIndex};
+use schemars::JsonSchema;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::marker::PhantomData;
 
 // version info for migration info
 const CONTRACT_NAME: &str = "crates.io:cw721-base";
 const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
 
-#[cfg_attr(not(feature = "library"), entry_point)]
-pub fn instantiate(
-    deps: DepsMut,
-    _env: Env,
-    _info: MessageInfo,
-    msg: InstantiateMsg,
-) -> StdResult<Response> {
-    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
-
-    let info = ContractInfoResponse {
-        name: msg.name,
-        symbol: msg.symbol,
-    };
-    CONTRACT_INFO.save(deps.storage, &info)?;
-    let minter = deps.api.addr_validate(&msg.minter)?;
-    MINTER.save(deps.storage, &minter)?;
-    Ok(Response::default())
+const DEFAULT_LIMIT: u32 = 10;
+const MAX_LIMIT: u32 = 30;
+
+pub struct Cw721Contract<'a, T, C>
```

Comment:
I would consider moving this to `state.rs` as it would be more consistent with way we stored it before, and leave only the methods implementation here - but I am not sure.

Reply:
I was thinking of that, but afaik, I need to implement all methods on a struct in the same module as the struct. Can I implement them in other modules in the same package?

Reply:
No, you need to do it on the same crate (and you need to have visibility for fields so they have to be `pub` or at least `pub(in: ...)`.

Reply:
Interesting.
I will try to pull this out later once we sort out the trait issues